### PR TITLE
Refactor and migrate `FIM` ITs

### LINKED ISSUE #17881: Refactor and migrate `FIM` integration tests from QA repository to Core repository
|Related Objective   |
|-----------------------|
|  wazuh/wazuh#17427             |

## Description

As part of wazuh/wazuh#17427, we need to refactor and migrate all legacy integration tests from the QA repository to the Core repository.

This issue aims to do this with the `FIM` tests. The following files are involved:

```
test_fim/
├───test_files
│   ├───test_ambiguous_confs
│   │   └───data
│   │       ├───configuration_templates
│   │       └───test_cases
│   ├───test_audit
│   │   └───data
│   │       ├───configuration_template
│   │       └───test_cases
│   ├───test_basic_usage
│   │   └───data
│   ├───test_env_variables
│   │   └───data
│   ├───test_file_limit
│   │   └───data
│   ├───test_follow_symbolic_link
│   │   └───data
│   ├───test_ignore
│   │   └───data
│   ├───test_inotify
│   │   └───data
│   ├───test_max_eps
│   │   └───data
│   ├───test_moving_files
│   │   └───data
│   ├───test_prefilter_cmd
│   │   └───data
│   ├───test_process_priority
│   │   └───data
│   ├───test_recursion_level
│   │   └───data
│   ├───test_report_changes
│   │   └───data
│   │       ├───configuration_template
│   │       └───test_cases
│   ├───test_restrict
│   │   └───data
│   ├───test_scan
│   │   └───data
│   ├───test_skip
│   │   └───data
│   ├───test_tags
│   │   └───data
│   ├───test_timezone_changes
│   │   └───data
│   ├───test_whodata_policy_change
│   │   └───data
│   │       ├───configuration_template
│   │       └───test_cases
│   ├───test_windows_audit_interval
│   │   └───data
│   └───test_windows_system_folder_redirection
│       └───data
│           ├───configuration_template
│           └───test_cases
├───test_registry
│   ├───test_registry_ambiguous_confs
│   │   └───data
│   ├───test_registry_basic_usage
│   │   └───data
│   ├───test_registry_checks
│   │   └───data
│   ├───test_registry_ignore
│   │   └───data
│   ├───test_registry_limit
│   │   └───data
│   ├───test_registry_multiple_registries
│   │   └───data
│   ├───test_registry_nodiff
│   │   └───data
│   ├───test_registry_report_changes
│   │   ├───data
│   │   └───test_disk_quota
│   │       └───data
│   ├───test_registry_restrict
│   │   └───data
│   └───test_registry_tags
│       └───data
└───test_synchronization
    └───data
        ├───configuration_template
        └───test_cases
```

### Tasks:

- [x] Migrate tests modules applying new format:
  - [x] test_files
      - [x] test_ambiguous_confs
          - [x] test_ignore_works_over_restrict.py
          - [x] test_whodata_ambiguous_thread.py
          - [x] test_whodata_works_over_realtime.py
      - [x] test_audit
          - [x] test_remove_audit.py
      - [x] test_basic_usage
          - [x] test_create_after_delete_dir.py (Cases with fim_mode = whodata and platform = winagent will be skipped due to unstable behavior on actions)
          - [x] test_delete_directory.py
          - [x] test_delete_hardlink_symlink.py
          - [x] test_delete_multiple_files.py
          - [x] test_empty_directories_tag.py
          - [x] test_fim_disabled.py
          - [x] test_move.py
          - [x] test_rename.py
      - [x] test_env_variables -> **These will be system tests.**
      - [x] test_checkers
          - [x] test_checkers.py -> windows cases remain to be migrated
      - [x] test_file_limit
          - [x] test_fill_capacity.py
          - [x] test_full_capacity_create_delete.py
          - [x] test_limit_disabled.py
      - [x] test_follow_symbolic_link
          - [x] test_audit_rules_with_symlink.py
          - [x] test_change_target.py
          - [x] test_cud.py
          - [x] test_disabled.py
          - [x] test_new_target_populated.py
      - [x] test_ignore
          - [x] test_ignore_valid.py (only for linux, windows test are skipped due to https://github.com/wazuh/wazuh/issues/9298)
      - [x] test_inotify
          - [x] test_inotify.py
      - [x] test_max_eps 
          - [x] test_max_eps.py (Cases with fim_mode = whodata and platform = winagent will be skipped due to unstable behavior on actions)
          - [x] test_sync_max_eps_scheduled.py -> **These will be system tests.**
      - [x] [test_moving_files](https://github.com/wazuh/wazuh/issues/22757)
          - [x] test_moving_files.py
      - [x] test_prefilter_cmd -> **These will be system tests.**
      - [x] [test_process_priority](https://github.com/wazuh/wazuh/issues/22758)
          - [x] test_process_priority.py
      - [x] [test_recursion_level](https://github.com/wazuh/wazuh/issues/22759)
          - [x] test_recursion_level.py
      - [x] [test_report_changes](https://github.com/wazuh/wazuh/issues/22760)
          - [x] test_diff_size_limit_configured.py
          - [x] test_diff_size_limit_default.py
          - [x] test_disk_quota_default.py
          - [x] test_disk_quota_disabled.py
          - [x] test_file_size_default.py
          - [x] test_file_size_disabled.py
          - [x] test_file_size_values.py
          - [x] test_large_changes.py
          - [x] test_report_changes_and_diff.py
          - [x] test_report_deleted_diff.py
      - [x] [test_restrict](https://github.com/wazuh/wazuh/issues/22761)
          - [x] test_restrict_valid.py
      - [x] test_scan -> **These will be system tests.**
      - [x] test_skip -> **These will be system tests.**
      - [x] [test_tags](https://github.com/wazuh/wazuh/issues/22762)
          - [x] test_tags.py
      - [x] test_timezone_changes -> **These will be system tests.**
      - [x] test_whodata_policy_change -> **These will be system tests.**
      - [x] [test_windows_audit_interval](https://github.com/wazuh/wazuh/issues/22763)
          - [x] test_windows_audit_interval.py
      - [x] [test_windows_system_folder_redirection](https://github.com/wazuh/wazuh/issues/22764)
         - [x] test_windows_system_folder_redirection.py
  - [x] test_registry
       - [x] [test_registry_ambiguous_confs](https://github.com/wazuh/wazuh/issues/22765)
           - [x] test_registry_ambiguous_complex.py
           - [x] test_registry_ambiguous_duplicated_entries.py
           - [x] test_registry_ambiguous_simple.py
       - [x] [test_registry_basic_usage](https://github.com/wazuh/wazuh/issues/22766) -> **Skipped tests, won't be migrated now.**
           - [x] test_basic_usage_delete_registry.py
           - [x] test_basic_usage_entries_match_key_count.py
           - [x] test_basic_usage_registry_baseline_generation.py
           - [x] test_basic_usage_registry_duplicated_entries.py
           - [x] test_basic_usage_registry_new_key.py
           - [x] test_long_registry_path.py
       - [x] [test_registry_checks](https://github.com/wazuh/wazuh/issues/22767)
           - [x] test_registry_checkers.py
       - [x] [test_registry_ignore](https://github.com/wazuh/wazuh/issues/22768) -> **Skipped tests, won't be migrated now.**
           - [x] test_ignore_registry.py
       - [x] [test_registry_limit](https://github.com/wazuh/wazuh/issues/22769) -> **Skipped tests, won't be migrated now.**
           - [x] test_registry_key_limit_full.py
           - [x] test_registry_key_limit_values.py
           - [x] test_registry_limit_capacity_alerts.py
           - [x] test_registry_limit_values.py
           - [x] test_registry_value_limit_full.py
       - [x] [test_registry_multiple_registries](https://github.com/wazuh/wazuh/issues/22770)
           - [x] test_multiple_keys.py
           - [x] test_multiple_registry_entries.py
       - [x] [test_registry_nodiff](https://github.com/wazuh/wazuh/issues/22771)
           - [x] test_registry_no_diff.py
       - [x] [test_registry_report_changes](https://github.com/wazuh/wazuh/issues/22772) -> **Skipped tests, won't be migrated now.**
           - [x] test_disk_quota
                - [x] test_registry_disk_quota_bigger_file_limit.py
	            - [x] test_registry_disk_quota_values.py
           - [x] test_registry_all_limits_disabled.py
           - [x] test_registry_diff_size_limit_values.py
           - [x] test_registry_file_size_values.py
           - [x] test_registry_report_changes_more_changes.py
           - [x] test_registry_report_changes.py
       - [x] [test_registry_restrict](https://github.com/wazuh/wazuh/issues/22773)
           - [x] test_registry_restrict.py
       - [x] [test_registry_tags](https://github.com/wazuh/wazuh/issues/22775)
           - [x] test_registry_tags.py
       - [x] [test_registry_wildcards](https://github.com/wazuh/wazuh/issues/22776)
           - [x] test_registry_wildcards.py
  - [x]  test_synchronization -> **These will be system tests.**

Framework branch: [17881_fim_dependencies_and_tests](https://github.com/wazuh/qa-integration-framework/tree/17881_fim_dependencies_and_tests)
Tests branch: [17881_fim_dependencies_and_tests](https://github.com/wazuh/wazuh/tree/17881_fim_dependencies_and_tests)

### PR DESCRIPTION
**Pull request by QU3B1M**
|Related issue|
|---|
| #17881|

## Description

This PR applies the refactored `FIM` Integration tests scenarios from the `/wazuh/wazuh-qa` repo. It also adds all the fixtures required for this test.

closes #17881

### CODE REVIEW

File: tests/integration/test_fim/conftest.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+import os
```
Comment (TomasTurina) [inline tests/integration/test_fim/conftest.py:1]: Missing header. Check all files with this problem.

File: tests/integration/test_fim/conftest.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+@pytest.fixture()
+def start_monitoring() -> None:
+    FileMonitor(WAZUH_LOG_PATH).start(generate_callback(MONITORING_PATH))
+
+
+@pytest.fixture(scope='module', autouse=True)
+def set_agent_config(request: pytest.FixtureRequest):
+    if not hasattr(request.module, 'test_configuration'):
+        return
+    if get_service() is WAZUH_MANAGER:
```
Comment (TomasTurina) [inline tests/integration/test_fim/conftest.py:74]: We should think if we want this. I think these tests should only run on agents, not in manager.

File: tests/integration/test_fim/test_files/test_ambiguous_confs/data/test_cases/cases_whodata_works_over_realtime.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+- name: whodata_thread_started_with_realtime_disabled
+  description: Check Whodata thread is enabled and is used with realtime disabled on same tag
+  configuration_parameters:
+    WHODATA: 'yes'
+    REALTIME: 'no'
+    TEST_DIR: !!python/object/apply:os.path.join [/, test_dir]
+  metadata:
+    folder_to_monitor: !!python/object/apply:os.path.join [/, test_dir]
+    file_to_monitor: !!python/object/apply:os.path.join [/, test_dir, test_file.log]
```
Comment (TomasTurina) [inline tests/integration/test_fim/test_files/test_ambiguous_confs/data/test_cases/cases_whodata_works_over_realtime.yaml:19]: Missing line a the end of the file. Check all files with this problem.

File: tests/integration/test_fim/test_files/test_audit/conftest.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Fixtures
```
Comment (TomasTurina) [inline tests/integration/test_fim/test_files/test_audit/conftest.py:1]: Ditto. Check all files and fix headers.

File: tests/integration/test_fim/test_files/test_audit/conftest.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import distro
+import re
+import pytest
+import subprocess
+
+
+@pytest.fixture(scope='module')
+def uninstall_audit():
```
Comment (TomasTurina) [inline tests/integration/test_fim/test_files/test_audit/conftest.py:9]: Maybe we can make install and uninstall `audit` functions in the framework?

File: tests/integration/test_fim/test_files/test_audit/conftest.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    """Uninstall auditd before test and install after test"""
+
+    # Check distro
+    linux_distro = distro.id()
+
+    if re.match(linux_distro, "centos"):
```
Comment (TomasTurina) [inline tests/integration/test_fim/test_files/test_audit/conftest.py:15]: Try to use platforms from framework or create new ones.

File: tests/integration/test_fim/test_files/test_basic_usage/conftest.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import pytest
```
Comment (TomasTurina) [inline tests/integration/test_fim/test_files/test_basic_usage/conftest.py:1]: Same.

File: tests/integration/test_fim/test_files/test_follow_symbolic_link/conftest.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import os
```
Comment (TomasTurina) [inline tests/integration/test_fim/test_files/test_follow_symbolic_link/conftest.py:1]: Same.

File: tests/integration/conftest.py [hunk 9]
```diff
@@ -638,3 +636,23 @@ def autostart_simulators(request: pytest.FixtureRequest) -> None:
...
+
+@pytest.fixture(scope='module')
+def set_environment_variables(request: pytest.FixtureRequest) -> None:
+    """
+    Create environment variables
+    """
+    if hasattr(request.module, 'environment_variables'):
+        environment_variables = getattr(request.module, 'environment_variables')
+        for env, value in environment_variables:
+            if sys.platform == 'win32':
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply tests/integration/conftest.py:675]: Done

File: tests/integration/test_fim/conftest.py (hunk 1 above)
Comment (sdvendramini) [reply tests/integration/test_fim/conftest.py:1]: Done

File: tests/integration/test_fim/test_files/test_ambiguous_confs/data/test_cases/cases_whodata_works_over_realtime.yaml (hunk 3 above)
Comment (sdvendramini) [reply tests/integration/test_fim/test_files/test_ambiguous_confs/data/test_cases/cases_whodata_works_over_realtime.yaml:19]: Done

File: tests/integration/test_fim/test_files/test_audit/conftest.py (hunk 5 above)
Comment (TomasTurina) [reply tests/integration/test_fim/test_files/test_audit/conftest.py:15]: Will do in another PR.

File: tests/integration/test_fim/test_files/test_audit/conftest.py (hunk 6 above)
Comment (TomasTurina) [reply tests/integration/test_fim/test_files/test_audit/conftest.py:21]: Will do in another PR.

**Review by TomasTurina — APPROVED** [review]

**Pull request merged**